File: packages/core/src/spec.ts
/**
 * FlowInk spec — the framework-agnostic description of an animated flow diagram.
 *
 * A diagram is a title, an optional subtitle, a set of manually positioned nodes,
 * and edges between them. Coordinates are absolute pixels on the canvas
 * (default 1200 × 640). Manual placement is a deliberate v1 decision: the diagrams
 * this library generalizes are small (≈10 nodes), and hand-positioned labels read
 * better than any auto-layout at this size.
 */

/** Semantic edge colors. The legend is part of the design language:
 *  sky = primary request path, emerald = auth, amber = secondary integration,
 *  rose = fallback/error path. */
export type FlowColor = 'sky' | 'emerald' | 'amber' | 'rose';

/** One box on the canvas. */
export interface FlowNode {
  /** Stable id referenced by edges. */
  id: string;
  /** Bold first line (node title). */
  label: string;
  /** Additional detail lines rendered smaller below the label. */
  lines?: string[];
  /** Top-left corner in canvas pixels. */
  x: number;
  y: number;
  /** Box size; defaults are derived from content when omitted. */
  width?: number;
  height?: number;
  /** Border "breathing" pulse. `true` uses the default 3s cycle;
   *  a number sets the duration in milliseconds. */
  pulse?: boolean | number;
}

/** A flow between two nodes. */
export interface FlowEdge {
  from: string;
  to: string;
  /** Small label rendered near the path midpoint. */
  label?: string;
  color?: FlowColor;
  /** `forward` animates left→right/top→bottom, `backward` the reverse,
   *  `none` draws a static line. Default `forward`. */
  direction?: 'forward' | 'backward' | 'none';
  /** Animate a small dot riding the path (CSS Motion Path). */
  packet?: boolean;
  /**
   * Manual SVG path data drawn from the `from` anchor to the `to` anchor.
   * When omitted, the renderer draws a simple line between the nearest
   * facing edges of the two boxes, curved slightly when they are not
   * horizontally aligned.
   */
  path?: string;
}

export interface FlowSpec {
  title: string;
  subtitle?: string;
  /** Canvas size, default 1200 × 640. */
  width?: number;
  height?: number;
  /** `dark` (default) is the showcase theme; `light` inverts to a paper look. */
  theme?: 'dark' | 'light';
  /** Corner chip text, e.g. a one-line product guarantee. */
  chip?: string;
  nodes: FlowNode[];
  edges: FlowEdge[];
}

File: dotnet/src/FlowInk.Core/Spec.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace FlowInk.Core;

/// <summary>Root FlowSpec document — mirrors the TypeScript schema exactly.</summary>
public sealed class FlowSpec
{
    [JsonPropertyName("title")] public string Title { get; set; } = "";
    [JsonPropertyName("subtitle")] public string? Subtitle { get; set; }
    [JsonPropertyName("width")] public int? Width { get; set; }
    [JsonPropertyName("height")] public int? Height { get; set; }
    [JsonPropertyName("theme")] public string? Theme { get; set; }
    [JsonPropertyName("chip")] public string? Chip { get; set; }
    [JsonPropertyName("nodes")] public List<FlowNode> Nodes { get; set; } = [];
    [JsonPropertyName("edges")] public List<FlowEdge> Edges { get; set; } = [];
}

public sealed class FlowNode
{
    [JsonPropertyName("id")] public string Id { get; set; } = "";
    [JsonPropertyName("label")] public string Label { get; set; } = "";
    [JsonPropertyName("lines")] public List<string>? Lines { get; set; }
    [JsonPropertyName("x")] public int X { get; set; }
    [JsonPropertyName("y")] public int Y { get; set; }
    [JsonPropertyName("width")] public int? Width { get; set; }
    [JsonPropertyName("height")] public int? Height { get; set; }
    [JsonPropertyName("pulse")] public JsonElement? Pulse { get; set; }

    public bool PulseIsBool(out bool value)
    {
        value = false;
        return Pulse is { ValueKind: JsonValueKind.True or JsonValueKind.False } &&
               (value = Pulse.Value.GetBoolean());
    }

    public int? PulseDurationMs() =>
        Pulse is { ValueKind: JsonValueKind.Number } ? (int)Pulse.Value.GetDouble() : null;
}

public sealed class FlowEdge
{
    [JsonPropertyName("from")] public string From { get; set; } = "";
    [JsonPropertyName("to")] public string To { get; set; } = "";
    [JsonPropertyName("label")] public string? Label { get; set; }
    [JsonPropertyName("color")] public string? Color { get; set; }
    [JsonPropertyName("direction")] public string? Direction { get; set; }
    [JsonPropertyName("packet")] public bool Packet { get; set; }
    [JsonPropertyName("path")] public string? Path { get; set; }
}
